### DISCUSSION TITLE
Enhancing Annotated Dependency

### DISCUSSION THREAD

**Discussion by ahmedabdou14** [fastapi/fastapi#9386] [Questions]
### First Check

- [x] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
""" Annotated Dependency """
dbDep = Annotated[AsyncSession, Depends(get_db)]
baseRepoDep = Annotated[BaseRepo, Depends(get_repo(BaseRepo))]
fakeRepoDep = Annotated[FakeRepo, Depends(get_repo(FakeRepo))]


router = APIRouter()

@router.get("/1")
async def root(repo: baseRepoDep):
    return {"message": "Hello World"}

@router.get("/2")
async def root(repo: fakeRepoDep):
    return {"message": "Hello World"}
```

### Description

Is there any way @tiangolo, we can have an extra feature for Annotated Dependency where we can specify the type of the repository we want to initialize but inside the api function itself and not have to declare a new Deps annotated variable to call.

Say we have 10 different repositories, should we have to explicitly create 10 annotated variables for dependency injection?
Do you think it is technically possible to have 1 factory annotated method where it takes the repo type as an argument and map it to the actual repo?

### Operating System

Linux

### Operating System Details

Ubuntu 22

### FastAPI Version

0.95.0

### Python Version

3.11

**Comment by 8thgencore** [comment]
Yes, it is technically possible to have a factory annotated method that takes the repository type as an argument and maps it to the actual repository. This way, you would not need to explicitly create 10 annotated variables for dependency injection.

Here's an example of how you could implement such a factory method:

```python
from typing import Type

def get_repo(repo_type: Type[BaseRepo]) -> BaseRepo:
    if repo_type == BaseRepo:
        return BaseRepo()
    elif repo_type == FakeRepo:
        return FakeRepo()
    # add more conditions for other repository types here
    else:
        raise ValueError(f"Unknown repository type: {repo_type}")
```

**Reply by ahmedabdou14** [reply]
Well, I usually go with the following approach, but it still requires explicit creation of `x` number of annotated variables

 ```python
from typing import Annotated, AsyncGenerator, Callable, Type
from fastapi import Depends
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.exc import SQLAlchemyError
from app.db.repos.base import BaseRepo, FakeRepo

from app.db.session import AsyncSessionFactory


async def get_db() -> AsyncGenerator[AsyncSession, None]:
    """Yields a db session from the pool"""
    try:
        async with AsyncSessionFactory() as session:
            session: AsyncSession
            yield session
    except SQLAlchemyError as e:
        await session.rollback()
        raise e
    finally:
        await session.close()


def get_repo(repo_type: Type[BaseRepo]) -> Callable[[AsyncSession], Type[BaseRepo]]:
    def _get_repo(db: AsyncSession = Depends(get_db)) -> Type[BaseRepo]:
        return repo_type(db)
    return _get_repo

    
""" Annotated Dependency """
dbDep = Annotated[AsyncSession, Depends(get_db)]
baseRepoDep = Annotated[BaseRepo, Depends(get_repo(BaseRepo))]
fakeRepoDep = Annotated[FakeRepo, Depends(get_repo(FakeRepo))]
```

However I don't see how the factory method would work, I am not sure how can you pass the repo type into the annotated variable in the api arguments
